### PR TITLE
Add on device beam search

### PR DESCRIPTION
**Pull request by katalinic-gc**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/generation/on_device_generation.py [hunk 1]
```diff
@@ -12,75 +12,381 @@
...
+        flat_beam_indices = beam_indices.view(-1)
+        flat_beam_indices = flat_batch_indices + flat_beam_indices
+        gathered_tensor = torch.index_select(tensor, 0, flat_beam_indices)
+        return self._unflatten_beam_dim(gathered_tensor, new_num_beams)
+
+    def forward(self, input_ids: torch.Tensor, absolute_step: torch.Tensor, **kwargs) -> torch.Tensor:
+        # Workaround for generic slice assignment self.running_sequences[:, self.absolute_step] = tokens
+        assert input_ids.shape[-1] == 1
+        input_ids_mask = self.input_ids_mask.to(input_ids.device)
+        padded_input_ids = torch.nn.functional.pad(input_ids, (0, self.max_length - 1))
```
Comment (jimypbr) [inline optimum/graphcore/generation/on_device_generation.py:209]: used padding token id for padding value?

Comment (katalinic-gc) [reply optimum/graphcore/generation/on_device_generation.py:209]: the padding value is irrelevant I believe, it gets masked out when multiplied by the mask

File: optimum/graphcore/generation/on_device_generation.py [hunk 2]
```diff
@@ -12,75 +12,381 @@
...
+        topk_running_sequences = self._gather_beams(
+            running_sequences, topk_beam_indices, self.batch_size, self.num_beams, beams_to_keep
+        )
+        topk_ids = topk_indices % vocab_size
+        topk_ids = topk_ids.unsqueeze(-1).int()
+
+        topk_sequences = poptorch.dynamic_update(topk_running_sequences, topk_ids, 2, absolute_step, 1)
+
+        # 4. Check which sequences have ended
+        did_topk_just_finished = torch.index_select(topk_sequences, 2, absolute_step).squeeze(2) == self.eos_token_id
```
[the remark this answers is not in the document]
Comment (katalinic-gc) [reply optimum/graphcore/generation/on_device_generation.py:247]: I blame upstream :P
